### DISCUSSION TITLE
Example of a custom rule to compile .S files?

### DISCUSSION THREAD

**Discussion by ilg-ul** [mesonbuild/meson#10431] [Q&A] [answered]
I'm trying to figure out how to add the rules to compile `.S` files to a project. 

The toolchains that I use require only adding `-x assembler-with-cpp` to the `*-gcc` line, the rest is the same as for compiling `.c` files.

My current `meson.build` file for the library is:

```
micro_os_plus_devices_cortexa_dependency = declare_dependency(
  include_directories: include_directories(
    'include',
  ),
  sources: files(
    'src/startup/boot.S',
    'src/startup/initialize-hardware-early.c',
    'src/startup/initialize-hardware.c',
    'src/startup/initialize-interrupts-stack.cpp',
    'src/startup/terminate.cpp',

    'src/exception-handlers.cpp',
    'src/vectors.S',
  ),
  compile_args: [
  ],
  dependencies: [
    micro_os_plus_semihosting_dependency
  ]
)
```

Unfortunately this does not work, the `.S` file is processed as a regular `.c` file, which fails.

Can someone suggest how to update the above definitions to add `-x assembler-with-cpp` for the assembly files?

Please note that my library is actually only a dependency, not a static library, and if we add custom rules, they should be forwarded to the application level, where all dependencies are compiled.

Thank you,

Liviu

**Comment by xclaesse** [comment]
If you're using gcc/clang it should just work if you add .S files into the sources of a target. What error do you get if you do that?

**Reply by ilg-ul** [reply]
> What error do you get if you do that?

```
> meson compile -C build/qemu-aarch64-meson-debug
ninja: Entering directory `/Users/ilg/My Files/WKS Projects/micro-os-plus.github/xPacks/utils-lists-xpack.git/build/qemu-aarch64-meson-debug'
[1/56] Compiling C object platform-qemu-aarch64/sample-test.elf.p/.._top_build_qemu-aarch64-meson-debug_xpacks_micro-os-plus-devices-cortexa_src_startup_boot.S.o
FAILED: platform-qemu-aarch64/sample-test.elf.p/.._top_build_qemu-aarch64-meson-debug_xpacks_micro-os-plus-devices-cortexa_src_startup_boot.S.o 
aarch64-none-elf-gcc -Iplatform-qemu-aarch64/sample-test.elf.p -Iplatform-qemu-aarch64 -I../../tests/platform-qemu-aarch64 -I../../tests/include -I../../tests/top/include -I../../tests/top/xpacks/micro-os-plus-diag-trace/include -I../../tests/platform-qemu-aarch64/include-platform -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-devices-qemu-cortexa/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/xpack-3rd-party-arm-cmsis-core-m/CMSIS/Core/Include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-devices-cortexa/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-semihosting/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-architecture-cortexa/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-startup/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-version/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-libs-c/include -I../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-libs-cpp/include -I../../tests/platform-qemu-aarch64/include-config -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Werror -std=c11 -g -fmessage-length=0 -fsigned-char -ffunction-sections -fdata-sections -fdiagnostics-color=always -DDEBUG -DMICRO_OS_PLUS_DEBUG -DMICRO_OS_PLUS_TRACE -DMICRO_OS_PLUS_INCLUDE_CONFIG_H -Wall -Waggregate-return -Wcast-align -Wcast-qual -Wconversion -Wdouble-promotion -Wduplicated-branches -Wduplicated-cond -Wextra -Wfloat-conversion -Wfloat-equal -Wformat-nonliteral -Wformat-overflow=2 -Wformat-security -Wformat-signedness -Wformat-truncation=2 -Wformat-y2k -Wformat=2 -Wlogical-op -Winvalid-pch -Wmissing-declarations -Wmissing-include-dirs -Wnull-dereference -Wpacked -Wpadded -Wpointer-arith -Wredundant-decls -Wshadow -Wshift-overflow=2 -Wsign-conversion -Wswitch-default -Wswitch-enum -Wundef -Wuninitialized -Wvla -Wbad-function-cast -Wc++-compat -Wduplicate-decl-specifier -Wmissing-prototypes -Wnested-externs -Wold-style-definition -Wstrict-prototypes -Warith-conversion -mcpu=cortex-a57 -mabi=lp64 -fno-move-loop-invariants -Werror -fno-omit-frame-pointer -D_POSIX_C_SOURCE=200809L -D__PROGRAM_START -DDEVICE_QEMU_CORTEX_A57 -DPLATFORM_QEMU_AARCH64 -MD -MQ platform-qemu-aarch64/sample-test.elf.p/.._top_build_qemu-aarch64-meson-debug_xpacks_micro-os-plus-devices-cortexa_src_startup_boot.S.o -MF platform-qemu-aarch64/sample-test.elf.p/.._top_build_qemu-aarch64-meson-debug_xpacks_micro-os-plus-devices-cortexa_src_startup_boot.S.o.d -o platform-qemu-aarch64/sample-test.elf.p/.._top_build_qemu-aarch64-meson-debug_xpacks_micro-os-plus-devices-cortexa_src_startup_boot.S.o -c ../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-devices-cortexa/src/startup/boot.S
../../tests/top/build/qemu-aarch64-meson-debug/xpacks/micro-os-plus-devices-cortexa/src/startup/boot.S:6:5: error: identifier "and" is a special operator name in C++ [-Werror=c++-compat]
    6 |     and    x0, x0, #3
      |     ^
cc1: all warnings being treated as errors
[14/56] Compiling C++ object platform-qemu-aarch64/sample-test.elf.p/.._src_sample-test.cpp.o
ninja: build stopped: subcommand failed.
error: running 'meson compile -C build/qemu-aarch64-meson-debug' failed
```

The cross compiler is based on the latest Arm bare-metal toolchains release (2022-02), which uses GCC 11.2, so it is relatively modern. 

> If you're using gcc/clang it should just work if you add .S files into the sources of a target.

It might be so, but I'm not aware of cross GCC automatically identifying `.S` files, I always passed `-x assembler-with-cpp` explicitly.

With CMake I pass this via `CMAKE_ASM_FLAGS`. Is there some way of passing flags to the assembler in meson? That would be easier than defining custom targets (which I have no experience with, yet).

**Comment by ilg-ul** [comment]
@eli-schwartz, any suggestions?

**Reply by eli-schwartz** [reply]
> It might be so, but I'm not aware of cross GCC automatically identifying `.S` files, I always passed `-x assembler-with-cpp` explicitly.
> 
> With CMake I pass this via `CMAKE_ASM_FLAGS`. Is there some way of passing flags to the assembler in meson? That would be easier than defining custom targets (which I have no experience with, yet).

Non-cross GCC definitely does it.

Meson's support for asm is pretty basic, all that happens is that the GNU compiler class defines .s as another file extension which CC knows how to compile. It's just treated as C sources as far as Meson knows. The flags used for the asm compilation are `c_args`.

**Comment by ilg-ul** [comment]
I also tried to define a separate dependency with the two assembly files, and pass it to micro_os_plus_devices_cortexa_dependency:

```
asm_dependency = declare_dependency(
  sources: files(
    'src/startup/boot.S',
    'src/vectors.S',
  ),
  compile_args: [
    '-x assembler-with-cpp'
  ],
)
```

This does not work either, because at the top level the compiler args are aggregated and all files are compiled with the `-x assembler-with-cpp` option.

**Comment by ilg-ul** [comment]
> Non-cross GCC definitely does it.

According to the GNU documentation, all modern GCC do it (which is nice, since `-x assembler-with-cpp` is no longer needed for `.S`), but that's not enough, the build system still needs to keep track of the assembly files in order to control the options passed via the command line, which are not necessarily identical to those of `.c`/`.cpp` files.

> Meson's support for asm is pretty basic ... The flags used for the asm compilation are `c_args`.

That's a pity, since the reason why my build failed is a consequence of not giving enough attention to the details, like not having fully separate, per language, definitions. 

In this specific case it is the lack of separate definitions for assembly files, but in other cases also the distinction between C and C++ options may be inadequate (like in the dependency definition, where there is a single `compile_args`, see above).

The reason why I had no problems building the same project with CMake is that the long list of warnings was not used by CMake in the assembly command line, and one of these warnings is harmful for assembly files (`-Wc++-compat`). The workaround for this meson build was to remove this warning from `c_args`, but this completely disables the compatibility checks for all C source files, which is not what it is desired.

To conclude, the ideal solution would be for meson to allow fully separate per language options, including for assembly, and to do this consistently everywhere, at the project level, in libraries, in dependencies, etc.

**Reply by eli-schwartz** [reply]
Meson also doesn't support the other dialects of asm, no one thinks this condition is optimal. :D There's quite a bit of discussion about improving this e.g. at #9379

**Comment by lbmeng** [comment]
As of meson v1.2.3, there is still no way to define separate compiler flags for assembly files. :(

**Reply by ilg-ul** [reply]
As there is still no way to define separate compiler flags for C/C++ when defining dependencies... :-(

**Reply by theoparis** [reply]
I just ran into this as well when building bionic libc with meson, as it needs the assembler-with-cpp flag to include the headers from .S files...

**Reply by dcbaker** [reply]
It's a known issue, I think we should solve it: https://github.com/mesonbuild/meson/issues/13940

This is very much a "patches welcome" situation, because while I think we all fundamentally agree that what we have works in some cases that it would be better to have it as a second language, but no one has immediate time to work on it.

**Comment by dcbaker** [comment, marked as answer]
There is an opened issue about this here: https://github.com/mesonbuild/meson/issues/13940

I'm locking the discussion because it seems that most of the comments here are really wanting to find that issue to comment on.
